text/0002-zmix-api/README.md

```diff
@@ -0,0 +1,178 @@
+
+Below is the proposed structure
+
+```
+libzmix
+   |
+   --- zmix_common
+          |
+          --- utils
+              errors
+       zmix_signatures
+          |
+          --- bbs
+              ps
+              cl
+              cdlnt
+       zmix_accumulators
+          |
+          --- rsa
+              hyperelliptic
+              cks
+              merkle
+       zmix_predicates
+           |
+           --- bulletproofs
+               snarks
+               r1cs
+       zmix_core
+           |
+           --- issuer
+               holder
+               verifier
+       Cargo.toml
+```
+
+`common` is used for traits and logic shared across 2 or more subcrates.
+
+`core` is organized according to the three roles of the zero-knowledge protocols.
+
+`signatures` are signatures that can be used to generate signature proofs of knowledge with selective disclosure proofs.
+Current signatures known to support this scheme are [BBS+](https://crypto.stanford.edu/~xb/crypto04a/groupsigs.pdf), [CL](https://groups.csail.mit.edu/cis/pubs/lysyanskaya/cl02b.pdfhttps://groups.csail.mit.edu/cis/pubs/lysyanskaya/cl02b.pdf),
+[PS](https://eprint.iacr.org/2015/525.pdf), [CDLNT](https://eprint.iacr.org/2020/016.pdf), and [delegatable credentials](https://acmccs.github.io/papers/p683-camenischA.pdf). There will be common code shared among signatures
+like interfaces and errors.
+
+`accumulators` are used for zero-knowledge set memberships. Current known schemes are accumulators of unknown order like 
+[RSA](https://eprint.iacr.org/2018/1188.pdf) and [Hyperelliptic Curves](https://eprint.iacr.org/2020/196.pdf), Elliptic Curve based like [CKS](https://link.springer.com/content/pdf/10.1007%2F978-3-642-00468-1_27.pdf), and [Merkle Tree](https://eprint.iacr.org/2019/1255.pdf) based.
+
+`predicates` are proofs for testing ranges and inequalities of [numeric](https://eprint.iacr.org/2017/1066.pdf) inputs or
+             other miscellaneous circuit based proofs like [PLONK](https://eprint.iacr.org/2019/953.pdf) or [SPARTAN](https://eprint.iacr.org/2019/550.pdf) and
+             [Verifiable Encryption](https://www.shoup.net/papers/verenc.pdf)
+
+# Drawbacks
+[drawbacks]: #drawbacks
+
+Creating the initial interfaces and APIs will be challenging and may slow down progress.
+
+# Rationale and alternatives
+[alternatives]: #alternatives
+
+The project could be organized by rust features alone. This has multiple drawbacks: cfg syntax sprinkled throughout the code,
+lots of combinations in Cargo.toml, and difficult to keep in mind all the various features when adding or modifying code. This makes 
+for a monolithic project that is consumable by users but requires documenting all features and complex combinations and coding
+for dangerous combinations.
+
+The project could be organized by just subcrates with no common API and consumers could pick and choose each one to use.
+The disadvantage is code duplication across projects.
```

Comment:
```suggestion
The project could be organized by rust features alone. 
This would have multiple drawbacks: 
- cfg syntax sprinkled throughout the code,
- lots of combinations in Cargo.toml, 
- the difficulty of keeping in mind all the various features when adding or modifying code. 

This would make for a monolithic project that is consumable by users but requires documenting all features 
and complex combinations and coding for dangerous combinations.

The project could be organized by just subcrates with no common API and consumers could pick and choose each one to use.
The disadvantage would be code duplication across projects and the lack of a common API.
```